controllers/vmware/virtualmachinegroup_reconciler.go

```diff
@@ -0,0 +1,400 @@
+)
+
+const (
+	// reconciliationDelay is the delay time for requeueAfter.
+	reconciliationDelay = 10 * time.Second
+	// ZoneAnnotationPrefix is the prefix used for placement decision annotations which will be set on VirtualMachineGroup.
+	ZoneAnnotationPrefix = "zone.cluster.x-k8s.io"
+)
+
+// VirtualMachineGroupReconciler reconciles VirtualMachineGroup.
+type VirtualMachineGroupReconciler struct {
+	Client   client.Client
+	Recorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters/status,verbs=get
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachinegroups,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachinegroups/status,verbs=get
+// +kubebuilder:rbac:groups=vmware.infrastructure.cluster.x-k8s.io,resources=vspheremachines,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinedeployments,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines,verbs=get;list;watch
+
+// This controller is introduced by CAPV to coordinate the creation and maintenance of
+// the VirtualMachineGroup (VMG) object with respect to the worker VSphereMachines in the Cluster.
+//
+// - Batch Coordination: Gating the initial creation of the VMG until all expected worker
+// VSphereMachines are present. This ensures the complete VM member list is sent to the VM
+// Service in a single batch operation due to a limitation of underlying service.
+//
+// - Placement Persistence: Persisting the MachineDeployment-to-Zone mapping (placement decision) as a
+// metadata annotation on the VMG object. This decision is crucial for guiding newer VMs created
+// during Day-2 operations such as scaling, upgrades, and remediations, ensuring consistency. This is also due to
+// a known limitation of underlying services.
+//
+// - Membership Maintenance: Dynamically updating the VMG's member list to reflect the current
+// state of VMs belonging to MachineDeployments (handling scale-up/down events).
+
+func (r *VirtualMachineGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the Cluster instance.
+	cluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	log = log.WithValues("Cluster", klog.KObj(cluster))
+
+	// If Cluster is deleted, just return as VirtualMachineGroup will be GCed and no extra processing needed.
+	if !cluster.DeletionTimestamp.IsZero() {
+		return reconcile.Result{}, nil
+	}
+
+	// If ControlPlane haven't initialized, requeue it since VSphereMachines of MachineDeployment will only be created after
+	// ControlPlane is initialized.
```

Comment:
I agree that we should wait, I am not sure if this statement is true (IIRC it is not)

Could be that VirtualMachine objects will only be created after CP initialized?